### PR TITLE
feat(experimental): add color picker experimental component

### PR DESCRIPTION
**Pull request by padanyuval**
add color picker experimental component

https://monday.monday.com/boards/3553578766/pulses/6933686896

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,10 +11,15 @@
...
     "typescript": "^4.7.3"
   },
   "workspaces": [
     "packages/*"
-  ]
+  ],
+  "dependencies": {
+    "react-color": "^2.19.3",
+    "react-colorful": "^5.6.1"
+  }
```
Comment (talkor) [inline package.json:24]: These should be in the `core` package's, this is the root package of the monorepo so deps here apply to all packages

File: package.json [hunk 2]
```diff
@@ -11,10 +11,15 @@
     "postinstall": "lerna run build --scope=vibe-storybook-components"
   },
   "devDependencies": {
+    "@types/react-color": "^3.0.12",
```
Comment (talkor) [inline package.json:14]: Same

File: packages/core/webpack/published-ts-components.js [hunk 3]
```diff
@@ -39,8 +39,8 @@ const publishedTSComponents = {
   Slider: "components/Slider/Slider",
   BreadcrumbsBar: "components/BreadcrumbsBar/BreadcrumbsBar",
   BreadcrumbItem: "components/BreadcrumbsBar/BreadcrumbItem/BreadcrumbItem",
-  ColorPicker: "components/ColorPicker/ColorPicker",
-  ColorPickerContent: "components/ColorPicker/components/ColorPickerContent/ColorPickerContent",
+  ColorPicker: "components/ColorPickerExperimental/ColorPickerExperimental",
+  ColorPickerContent: "components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent",
```
Comment (talkor) [inline packages/core/webpack/published-ts-components.js:43]: This shouldn't be changed as it break the existing ColorPicker, you only need the exports in next.ts

File: yarn.lock
Comment (talkor) [inline yarn.lock:1]: There seem to be a lot of changes here, can you please try to revert the changes, install the required packages in the core package json and make sure the changes here are intended?

File: packages/core/src/components/ColorPickerExperimental/ColorPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      customColorsLimit,
+      onCustomColorsSave = NOOP
+    },
+    ref
+  ) => {
+    const componentRef = useRef(null);
+    const mergedRef = useMergeRef(ref, componentRef);
+
+    const onChange = useCallback(onSave, [onSave]);
+    const onCustomColorsChange = useCallback(onCustomColorsSave, [onCustomColorsSave]);
```
Comment (talkor) [inline packages/core/src/components/ColorPickerExperimental/ColorPicker.tsx:98]: The useCallback can be already inside the ColorPickerContent

File: packages/core/src/components/ColorPickerExperimental/__tests__/ColorPicker.test.tsx
Comment (talkor) [inline packages/core/src/components/ColorPickerExperimental/__tests__/ColorPicker.test.tsx:1]: What about some tests for the custom functionality?

File: packages/core/src/components/ColorPickerExperimental/__tests__/ColorPicker.interactions.js
Comment (talkor) [inline packages/core/src/components/ColorPickerExperimental/__tests__/ColorPicker.interactions.js:1]: What about some tests for the custom functionality?

**Review by talkor — CHANGES_REQUESTED** [review]
Very cool! I feel like there's should be more focus on the separation of components and concerns, plus let's discuss the usage of the 3rd party libraries, but that's a great start

File: packages/core/src/components/ColorPickerExperimental/services/ColorPickerService.tsx
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/services/ColorPickerService.tsx:1]: moved this function to the existing colors-utils.ts file

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+import { ColorPickerColorsGrid } from "./ColorPickerColorsGrid";
+import { VibeComponentProps, VibeComponent, SubIcon, withStaticProps } from "../../../../types";
+import useMergeRef from "../../../../hooks/useMergeRef";
+import { TextType } from "../../../Text/TextConstants";
+import styles from "./ColorPickerContent.module.scss";
+import cx from "classnames";
+import Text from "../../../Text/Text";
+import { Button } from "../../../Button";
+import { Tooltip } from "../../../Tooltip";
+import GeneralColorPickerSection from "./GeneralColorPicker";
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:30]: changed it to "AdvancedColorPicker"

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      colorShape,
+      "data-testid": dataTestId,
+      isDisabled,
+      isEditing
+    },
+    _ref
+  ) => {
+    const isMondayColor = useMemo(() => (contentColors as readonly string[]).includes(color), [color]); // casting to any since color can be one of the system content colors but can also be a custom one
+    const colorAsStyle = isMondayColor ? ColorUtils.getMondayColorAsStyle(color, colorStyle) : color;
+    const disabledColor = isMondayColor ? ColorUtils.getMondayColorAsStyle(color, ColorStyle.SELECTED) : color;
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx:57]: the SELECTED style is acting like disabled (the regular color with brightness). 
don't know why it's called selected

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            }}
+            onClick={onClick}
+            tabIndex="-1"
+            onMouseDown={e => e.preventDefault()} // this is for quill to not lose the selection
+            disabled={isDisabled}
+          >
+            <div className={cx(styles.colorIndicatorWrapper)} style={colorIndicatorWrapperStyle}>
+              {shouldRenderIcon && (
+                <Icon
+                  icon={isEditing ? Edit : isSelected ? SelectedIndicatorIcon : ColorIndicatorIcon}
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx:129]: regarding the UX concern -
We did usability and all of the users started hovering over the colors and went to edit it. There is no issue in discoverability and understanding. It is the common and intuitive thing to do - hover over colors.
The second reason we didn't add it, is because it is not a consistent with other behaviors in the platform and in addition to that it is not a visually aesthetic solution

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+              background: isDisabled
+                ? disabledColor
+                : shouldRenderIndicatorWithoutBackground
+                ? "transparent"
+                : colorAsStyle
+            }}
+            onClick={onClick}
+            tabIndex="-1"
+            onMouseDown={e => e.preventDefault()} // this is for quill to not lose the selection
+            disabled={isDisabled}
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx:124]: removed

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.tsx:57]: removed it. should keep the same color

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.module.scss [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+.itemWrapper:hover:not(.active):not(.selectedColor):not(.disabled) .feedbackIndicator {
+  @include absolute-border(4px, primary-background-hover-color, -4px, 8px);
+}
+
+.itemWrapper .colorItem {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  border: 2px solid transparent;
+  border-radius: calc(4px * var(--border-radius-multiplier, 1));
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerItemComponent/ColorPickerItemComponent.module.scss:68]: returned it

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+  colorShape?: ColorShapes;
+  tooltipContentByColor?: Partial<Record<CONTENT_COLORS_VALUES, string>>;
+  noColorText?: string;
+  shouldRenderIndicatorWithoutBackground?: boolean;
+  isBlackListMode?: boolean;
+  numberOfColorsInLine?: number;
+  focusOnMount?: boolean;
+  isMultiselect?: boolean;
+  customColorsList?: string[];
+  customColorsLimit?: number;
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:51]: required feature

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+   * Used to force the component render the colorList prop as is. Usually, this flag should not be used. It's intended only for edge cases.
+   * Usually, only "monday colors" will be rendered (unless blacklist mode is used). This flag will override this behavior.
+   */
+  forceUseRawColorList?: boolean;
+  /**
+   * Used to enable color name tooltip on each color in the component. it's incompatible with forceUseRawColorList flag.
+   * When "tooltipContentByColor" is supplied, it will override the color name tooltip.
+   */
+  showColorNameTooltip?: boolean;
+  allowCustomColors?: boolean;
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:63]: in case we are not sending customColors but want to allow it (inital case)

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    const onClearButton = useCallback(() => {
+      onValueChange(null);
+    }, [onValueChange]);
+
+    const colorsToRender = useMemo(() => {
+      if (forceUseRawColorList) {
+        return colorsList;
+      }
+      if (allowCustomColors && isEmpty(colorsList)) {
+        return [...DEFAULT_COLORS];
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:118]: in contentUtils (contentColors) but we want to change the order they presented

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+        }
+        onValueChange(colors);
+      },
+      [isMultiselect, onValueChange, value]
+    );
+
+    const positions = useMemo(() => [{ topElement: colorsRef, bottomElement: buttonRef }], []);
+    const keyboardContext = useGridKeyboardNavigationContext(positions, gridRef);
+    const width = calculateColorPickerWidth(colorSize, numberOfColorsInLine);
+    const [shouldShowGeneralPicker, setShouldShowGeneralPicker] = useState(false);
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:146]: it controls the whole content should be displayed (the regular grid and the advanced picker)

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+    const DefaultColorsBar = () => (
+      <div className={cx(styles.defaultColorsBar)}>
+        <Text type={TextType.TEXT2} color={Text.colors.SECONDARY}>
+          {"Default colors"}
+        </Text>
+      </div>
+    );
+
+    const CustomColorsBar = () => {
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:210]: I don't agree.. it's part of the color picker content to show

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+            size={Button.sizes.XS}
+            kind={Button.kinds.SECONDARY}
+            disabled={customColors.length === 0}
+            onClick={toggleEditingMode}
+          >
+            Edit
+          </Button>
+        </Tooltip>
+      );
+      const doneButton = (
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.tsx:223]: removed

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 16]
```diff
@@ -0,0 +1,67 @@
+.customSketchPicker {
+  .sketch-picker {
+    box-shadow: none !important;
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:3]: that's the last resort :) all the children should not have this property, didn't find other solution

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    border-radius: 8px;
+    border: 1px solid #eaeaea;
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:5]: removed the sketch-picker css

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  align-items: center;
+  overflow: hidden;
+  justify-content: space-between;
+  padding-right: 4px;
+  padding-left: 4px;
+  margin-top: 16px;
+  margin-bottom: 4px;
+}
+
+.generalPickerOptionsBar {
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:54]: nice! changed it to use the Flex component

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    .saturation {
+      border-radius: 16px !important; /* Increase for more rounded edges */
+      background-color: white !important; /* Set to 'transparent' if you prefer */
+    }
+
+    .hue-horizontal {
+      border-radius: 8px !important; /* Adjust this value to control the border radius of the color bar */
+    }
+
+    .flexbox-fix {
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:24]: removed the sketch-picker css

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 20]
```diff
@@ -0,0 +1,67 @@
+.customSketchPicker {
+  .sketch-picker {
+    box-shadow: none !important;
+    border-radius: 8px;
+    border: 1px solid #eaeaea;
+
+    input {
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:7]: removed the sketch-picker css

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:1]: removed the sketch-picker css - i'm not using the sketch picker, forgot to delete it. sorry

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss [hunk 21]
```diff
@@ -0,0 +1,67 @@
+.customSketchPicker {
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerContent.module.scss:1]: you right - removed it

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerColorsGrid.tsx [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+      colorSize,
+      tooltipContentByColor,
+      colorShape,
+      showColorNameTooltip: showColorNameTooltip,
+      id,
+      "data-testid": dataTestId,
+      openGeneralPicker,
+      isEditing,
+      isDisabled,
+      allowCustomColors,
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerColorsGrid.tsx:75]: we need both

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerColorsGrid.tsx [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+        })}
+        {allowCustomColors && !isEditing && (
+          <Tooltip
+            content={
+              customColors && customColorsLimit && customColors.length >= customColorsLimit
+                ? "You've hit the custom colors limit"
+                : "Add custom color"
+            }
+          >
+            <IconButton
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/ColorPickerColorsGrid.tsx:129]: changed it

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/GeneralColorPicker.tsx [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+            <Button kind={Button.kinds.TERTIARY} size={Button.sizes.SMALL} onClick={closePicker}>
+              Cancel
+            </Button>
+            <Button size={Button.sizes.SMALL} onClick={handleEditChange}>
+              Done
+            </Button>
+          </div>
+        </div>
+      ) : (
+        <div className={styles.optionsBar}>
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/GeneralColorPicker.tsx:63]: added

File: packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/GeneralColorPicker.tsx [hunk 25]
```diff
@@ -0,0 +1,76 @@
+import React from "react";
+import { MaterialPicker } from "react-color";
+import { HexColorPicker } from "react-colorful";
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/components/ColorPickerContent/GeneralColorPicker.tsx:3]: moved to atmosphere :)

File: packages/core/src/components/ColorPickerExperimental/ColorPickerConstants.ts [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+  DELETE = "delete"
+}
+
+export const DEFAULT_COLOR = "#000000";
+
+export type ColorPickerValueOnly = CONTENT_COLORS_VALUES | string;
+export type ColorPickerArrayValueOnly = CONTENT_COLORS_VALUES[] | string[];
+export type ColorPickerValue = ColorPickerValueOnly | ColorPickerArrayValueOnly;
+
+export const DEFAULT_COLORS = [
```
[the remark this answers is not in the document]
Comment (padanyuval) [reply packages/core/src/components/ColorPickerExperimental/ColorPickerConstants.ts:32]: in contentUtils (contentColors) but we want to change the order they presented so we have them here
